File: server/src/post/PostDao.ts
import { asCamelCase } from '../common/formatters';
import { Db } from '../common/types';
import { Post } from '../../shared/types/post';

export class PostDao {
  private tableColumns = [
    'id',
    'created_at',
    'updated_at',
    'subject',
    'content',
    'author',
  ]
    .map(column => `posts.${column}`)
    .join(', ');

  public async getAllPosts({ db }: { db: Db }): Promise<Post[]> {
    const data = await db.any(
      `
        SELECT ${this.tableColumns}
        FROM posts
        ORDER BY created_at DESC
      `
    );

    return asCamelCase(data);
  }

  public async createPost({
    db,
    subject,
    content,
    author,
  }: {
    db: Db;
    subject: string;
    content: string;
    author: string;
  }) {
    const data = await db.one(
      `
        INSERT INTO posts (
          subject,
          content,
          author
        ) VALUES (
          $[subject],
          $[content],
          $[author]
        )
        RETURNING ${this.tableColumns}
      `,
      {
        subject,
        content,
        author,
      }
    );

    return asCamelCase(data);
  }
}

File: alternatives/server-flask/src/core/daos/post_dao.py
import dataclasses
from psycopg2 import sql
import typing
from src.common.setup import db
from ..types.post import DBPost, Post


DB_POST_FIELDS: sql.SQL = sql.SQL(",").join(
    sql.Identifier(f.name) for f in dataclasses.fields(DBPost)
)


class PostDao:
    def __init__(self, database=None):
        self._db = database or db

    def search(self) -> typing.List[DBPost]:
        """Search posts from database."""
        # TODO: handle search filters, paging and total count
        statement = sql.SQL(
            """
            SELECT      {fields}
            FROM        posts
            ORDER BY    created_at DESC
            """
        ).format(
            fields=DB_POST_FIELDS,
        )
        posts = self._db.execute(statement)
        return [DBPost(*post) for post in posts]

    def create(self, post: Post) -> DBPost:
        """Create a new post to database."""
        insert_fields = []
        placeholders = []
        for f in dataclasses.fields(post):
            insert_fields.append(sql.Identifier(f.name))
            placeholders.append(sql.Placeholder(f.name))

        statement = sql.SQL(
            """
            INSERT INTO posts ({insert_fields})
            VALUES      ({placeholders})
            RETURNING   {fields}
            """
        ).format(
            insert_fields=sql.SQL(",").join(insert_fields),
            placeholders=sql.SQL(",").join(placeholders),
            fields=DB_POST_FIELDS,
        )

        record = self._db.execute(
            statement,
            params=dataclasses.asdict(post),
            single=True,
        )
        return DBPost(*record)
